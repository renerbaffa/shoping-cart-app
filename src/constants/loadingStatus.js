export const FAILED = 'FAILED';
export const RETRIEVED = 'RETRIEVED';
export const RETRIEVING = 'RETRIEVING';
