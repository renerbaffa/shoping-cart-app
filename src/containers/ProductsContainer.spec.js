import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import ConnectedProductsContainer, { ProductsContainer } from './ProductsContainer';

import { GRID, LIST } from '../constants/ViewOptions';
import { RETRIEVING, RETRIEVED } from '../constants/loadingStatus';

jest.mock('../components/products/ProductsList', () => 'ProductsList');

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('<ProductsContainer />', () => {
  let defaultComponent;
  let onFetchProducts;
  let wrapper;

  beforeEach(() => {
    onFetchProducts = jest.fn();
    defaultComponent = (
      <ProductsContainer
        onFetchProducts={onFetchProducts}
      />
    );
    wrapper = shallow(defaultComponent);
  });

  it('should render correctly', () => {
    const switchViewOptions = renderer.create(defaultComponent).toJSON();
    expect(switchViewOptions).toMatchSnapshot();
  });

  it('should fetch projects when did mount', () => {
    onFetchProducts.mockReset();
    wrapper.instance().componentDidMount();
    expect(onFetchProducts).toHaveBeenCalled();
  });

  it('should update state with new view when handling change', () => {
    expect(wrapper.state().currentView).toBe(GRID);
    wrapper.instance().handleSwitchView(LIST);
    expect(wrapper.state().currentView).toBe(LIST);
  });

  describe('on component connect to redux isLoading flag', () => {
    it('should set isLoading as false when flag is not RETRIEVING', () => {
      const store = mockStore({
        communication: { areProjectsLoading: RETRIEVED },
      });

      wrapper = shallow(
        <ConnectedProductsContainer
          store={store}
          onFetchProducts={onFetchProducts}
        />
      );

      expect(wrapper.props().isLoading).toBeFalsy();
    });

    it('should set isLoading as true when flag is RETRIEVING', () => {
      const store = mockStore({
        communication: { areProjectsLoading: RETRIEVING },
      });

      wrapper = shallow(
        <ConnectedProductsContainer
          store={store}
          onFetchProducts={onFetchProducts}
        />
      );

      expect(wrapper.props().isLoading).toBeTruthy();
    });
  });
});

