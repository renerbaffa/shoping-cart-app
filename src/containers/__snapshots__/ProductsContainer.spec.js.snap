// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsContainer /> should render correctly 1`] = `
<div
  className="ProductsContainer-container"
>
  <div
    className="limited-width ProductsContainer-border"
  >
    <div
      className="space-between"
    >
      <div>
        <form
          className="SearchForm-form"
        >
          <input
            className="SearchForm-input"
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value=""
          />
          <button
            className="SearchForm-button Button"
            onClick={[Function]}
          >
            Search
          </button>
        </form>
      </div>
      <div>
        <button
          className="SwitchViewOptions-button Button Button-selected"
          onClick={[Function]}
        >
          Grid
        </button>
        <button
          className="SwitchViewOptions-button Button"
          onClick={[Function]}
        >
          List
        </button>
      </div>
    </div>
    <div
      className="ProductsContainer-content"
    >
      <ProductsList
        currentView="GRID"
      />
    </div>
  </div>
</div>
`;
